In [2]:
from transformers import pipeline

https://huggingface.co/distilbert/distilbert-base-uncased

In [3]:
mask_pipeline = pipeline("fill-mask", model="distilbert-base-uncased")

mask_pipeline

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [4]:
text = "He loves to play [MASK] on weekends."

text

'He loves to play [MASK] on weekends.'

In [5]:
mask_pipeline(text)

[{'score': 0.3326966464519501,
  'token': 5439,
  'token_str': 'golf',
  'sequence': 'he loves to play golf on weekends.'},
 {'score': 0.09304720163345337,
  'token': 11662,
  'token_str': 'poker',
  'sequence': 'he loves to play poker on weekends.'},
 {'score': 0.07858829200267792,
  'token': 5093,
  'token_str': 'tennis',
  'sequence': 'he loves to play tennis on weekends.'},
 {'score': 0.06728240102529526,
  'token': 3598,
  'token_str': 'baseball',
  'sequence': 'he loves to play baseball on weekends.'},
 {'score': 0.050590675324201584,
  'token': 7433,
  'token_str': 'chess',
  'sequence': 'he loves to play chess on weekends.'}]

In [6]:
text = "She found the book very [MASK]."

mask_pipeline(text)

[{'score': 0.09489227831363678,
  'token': 5875,
  'token_str': 'interesting',
  'sequence': 'she found the book very interesting.'},
 {'score': 0.059770409017801285,
  'token': 17160,
  'token_str': 'fascinating',
  'sequence': 'she found the book very fascinating.'},
 {'score': 0.05858148634433746,
  'token': 14036,
  'token_str': 'entertaining',
  'sequence': 'she found the book very entertaining.'},
 {'score': 0.04417254403233528,
  'token': 19142,
  'token_str': 'amusing',
  'sequence': 'she found the book very amusing.'},
 {'score': 0.04347850754857063,
  'token': 16004,
  'token_str': 'appealing',
  'sequence': 'she found the book very appealing.'}]

In [7]:
text = "The weather today is [MASK]."

mask_pipeline(text, top_k=3)

[{'score': 0.1265837550163269,
  'token': 10256,
  'token_str': 'mild',
  'sequence': 'the weather today is mild.'},
 {'score': 0.09278498589992523,
  'token': 24706,
  'token_str': 'cloudy',
  'sequence': 'the weather today is cloudy.'},
 {'score': 0.04285825416445732,
  'token': 24166,
  'token_str': 'stormy',
  'sequence': 'the weather today is stormy.'}]

In [8]:
text =  "She wore a [MASK] dress and a [MASK] hat."

mask_pipeline(text)

[[{'score': 0.07334267348051071,
   'token': 2317,
   'token_str': 'white',
   'sequence': '[CLS] she wore a white dress and a [MASK] hat. [SEP]'},
  {'score': 0.05665748193860054,
   'token': 2417,
   'token_str': 'red',
   'sequence': '[CLS] she wore a red dress and a [MASK] hat. [SEP]'},
  {'score': 0.04641922563314438,
   'token': 2304,
   'token_str': 'black',
   'sequence': '[CLS] she wore a black dress and a [MASK] hat. [SEP]'},
  {'score': 0.04326103627681732,
   'token': 2630,
   'token_str': 'blue',
   'sequence': '[CLS] she wore a blue dress and a [MASK] hat. [SEP]'},
  {'score': 0.038923703134059906,
   'token': 5030,
   'token_str': 'wedding',
   'sequence': '[CLS] she wore a wedding dress and a [MASK] hat. [SEP]'}],
 [{'score': 0.18764641880989075,
   'token': 13137,
   'token_str': 'straw',
   'sequence': '[CLS] she wore a [MASK] dress and a straw hat. [SEP]'},
  {'score': 0.15933655202388763,
   'token': 11762,
   'token_str': 'cowboy',
   'sequence': '[CLS] she wore a 

In [9]:
text = "On vacation, I like to explore the [MASK] and try a new [MASK]."

mask_pipeline(text)

On vacation, I like to explore the [MASK] and try a new [MASK].


[[{'score': 0.05310157686471939,
   'token': 10833,
   'token_str': 'countryside',
   'sequence': '[CLS] on vacation, i like to explore the countryside and try a new [MASK]. [SEP]'},
  {'score': 0.03610825911164284,
   'token': 4020,
   'token_str': 'mountains',
   'sequence': '[CLS] on vacation, i like to explore the mountains and try a new [MASK]. [SEP]'},
  {'score': 0.03372696787118912,
   'token': 4153,
   'token_str': 'ocean',
   'sequence': '[CLS] on vacation, i like to explore the ocean and try a new [MASK]. [SEP]'},
  {'score': 0.031179824844002724,
   'token': 19350,
   'token_str': 'outdoors',
   'sequence': '[CLS] on vacation, i like to explore the outdoors and try a new [MASK]. [SEP]'},
  {'score': 0.029092462733387947,
   'token': 3509,
   'token_str': 'beach',
   'sequence': '[CLS] on vacation, i like to explore the beach and try a new [MASK]. [SEP]'}],
 [{'score': 0.05544554814696312,
   'token': 6172,
   'token_str': 'adventure',
   'sequence': '[CLS] on vacation, i li

Even if the training data used for this model could be characterized as fairly neutral, this model can have biased predictions

In [10]:
text = "The gentleman worked as a [MASK]."

mask_pipeline(text)

[{'score': 0.12069946527481079,
  'token': 10533,
  'token_str': 'carpenter',
  'sequence': 'the gentleman worked as a carpenter.'},
 {'score': 0.09845925122499466,
  'token': 20987,
  'token_str': 'blacksmith',
  'sequence': 'the gentleman worked as a blacksmith.'},
 {'score': 0.06329615414142609,
  'token': 22701,
  'token_str': 'tailor',
  'sequence': 'the gentleman worked as a tailor.'},
 {'score': 0.05579977110028267,
  'token': 7500,
  'token_str': 'farmer',
  'sequence': 'the gentleman worked as a farmer.'},
 {'score': 0.045212168246507645,
  'token': 7805,
  'token_str': 'clerk',
  'sequence': 'the gentleman worked as a clerk.'}]

In [11]:
text = "The lady worked as a [MASK]."

mask_pipeline(text)

[{'score': 0.17885352671146393,
  'token': 10850,
  'token_str': 'maid',
  'sequence': 'the lady worked as a maid.'},
 {'score': 0.08888010680675507,
  'token': 6821,
  'token_str': 'nurse',
  'sequence': 'the lady worked as a nurse.'},
 {'score': 0.07091602683067322,
  'token': 22583,
  'token_str': 'housekeeper',
  'sequence': 'the lady worked as a housekeeper.'},
 {'score': 0.04480186477303505,
  'token': 5660,
  'token_str': 'cook',
  'sequence': 'the lady worked as a cook.'},
 {'score': 0.03941650688648224,
  'token': 10533,
  'token_str': 'carpenter',
  'sequence': 'the lady worked as a carpenter.'}]

* https://huggingface.co/docs/transformers/model_doc/auto
* https://huggingface.co/bert-base-uncased

In [14]:
import tensorflow as tf

from transformers import TFAutoModelForMaskedLM, AutoTokenizer

In [15]:
BERT_MODEL = "bert-base-uncased"

BERT_MODEL

'bert-base-uncased'

In [16]:
tokenizer = AutoTokenizer.from_pretrained(BERT_MODEL)

tokenizer

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

BertTokenizerFast(name_or_path='bert-base-uncased', vocab_size=30522, model_max_length=512, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}

In [17]:
model = TFAutoModelForMaskedLM.from_pretrained(BERT_MODEL)

model

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

All PyTorch model weights were used when initializing TFBertForMaskedLM.

All the weights of TFBertForMaskedLM were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFBertForMaskedLM for predictions without further training.



- Run this with the following sentences

` She wore an elegant [MASK] to the event.`

`He wore a elegant [MASK] to the event.`

In [18]:
text = "He wore an elegant [MASK] to the event."

text

'He wore an elegant [MASK] to the event.'

In [19]:
input_ids = tokenizer.encode(text, return_tensors="tf")

input_ids

<tf.Tensor: shape=(1, 11), dtype=int32, numpy=
array([[  101,  2002,  5078,  2019, 11552,   103,  2000,  1996,  2724,
         1012,   102]], dtype=int32)>

##### finds the indices of masked tokens in the input_ids (there is just one token_id = 103)

In [20]:
mask_token_indices = tf.where(input_ids == tokenizer.mask_token_id)

mask_token_indices

<tf.Tensor: shape=(1, 2), dtype=int64, numpy=array([[0, 5]])>

#####  iterates over the indices of masked tokens in the input obtains the model predictions for each masked token and appends the decoded word corresponding to the highest probability token to predicted_words

In [24]:
# Iterate over all the masks present in the input
predicted_words = []

# Iterate over all the masks present in the input
for i in range(mask_token_indices.shape[0]):
    mask_token_index = mask_token_indices[i, 1]

    predictions = model(input_ids)

    # Get logits for the masked word prediction
    # Logits shape [batch_size, length_of_input, num_predicted_words]
    predicted_token_logits = predictions.logits[0, mask_token_index, :]

    # Apply softmax to get probabilities
    predicted_token_probabilities = tf.nn.softmax(predicted_token_logits, axis=-1)

    # Find the predicted token id with the highest probability
    predicted_token_ids = tf.argmax(predicted_token_probabilities, axis=-1)

    # Decode the predicted token id to a word
    predicted_word = tokenizer.decode([predicted_token_ids.numpy()])

    # Check if the prediction is not null or empty
    if predicted_word and predicted_word.strip():
        predicted_words.append(predicted_word)
    else:
        predicted_words.append("[MASK]")  # or some other placeholder for null prediction


In [25]:
predicted_words

['suit']

In [26]:
for predicted_word in predicted_words:
    text = text.replace("[MASK]", predicted_word, 1)

print(text)

He wore an elegant suit to the event.
